src/torchcodec/_core/DeviceInterface.h

```diff
@@ -40,20 +64,77 @@ class DeviceInterface {
       FrameOutput& frameOutput,
       std::optional<torch::Tensor> preAllocatedOutputTensor = std::nullopt) = 0;
 
+  // ------------------------------------------
+  // Extension points for custom decoding paths
+  // ------------------------------------------
+
+  // Override to return true if this device interface can decode packets
+  // directly
+  virtual bool canDecodePacketDirectly() const {
+    return false;
+  }
+
+  // Moral equivalent of avcodec_send_packet()
+  // Returns AVSUCCESS on success, AVERROR(EAGAIN) if decoder queue full, or
+  // other AVERROR on failure
+  virtual int sendPacket([[maybe_unused]] ReferenceAVPacket& avPacket) {
+    TORCH_CHECK(
+        false,
+        "Send/receive packet decoding not implemented for this device interface");
+    return AVERROR(ENOSYS);
+  }
+
+  // Moral equivalent of avcodec_receive_frame()
+  // Returns AVSUCCESS on success, AVERROR(EAGAIN) if no frame ready,
+  // AVERROR_EOF if end of stream, or other AVERROR on failure
+  virtual int receiveFrame(
+      [[maybe_unused]] UniqueAVFrame& avFrame,
+      [[maybe_unused]] int64_t desiredPts) {
+    TORCH_CHECK(
+        false,
+        "Send/receive packet decoding not implemented for this device interface");
+    return AVERROR(ENOSYS);
+  }
+
+  // Flush remaining frames from decoder
+  virtual void flush() {
+    // Default implementation is no-op for standard decoders
+    // Custom decoders can override this method
+  }
+
+  // Apply bitstream filter if needed, returns pointer to packet to use
+  // Default implementation returns the original packet (no filtering)
+  virtual ReferenceAVPacket* applyBSF(
+      ReferenceAVPacket& packet,
+      [[maybe_unused]] AutoAVPacket& filteredAutoPacket,
+      [[maybe_unused]] ReferenceAVPacket& filteredPacket) {
+    return &packet; // No filtering by default
+  }
+
  protected:
   torch::Device device_;
 };
 
 using CreateDeviceInterfaceFn =
     std::function<DeviceInterface*(const torch::Device& device)>;
 
+bool registerDeviceInterface(
+    const DeviceInterfaceKey& key,
+    const CreateDeviceInterfaceFn createInterface);
+
+// Backward-compatible registration function when variant is "default"
+// TODONVDEC P2 We only need this if someone in the wild has already started
+// registering their own interfaces. Ask Dmitry.
```

Comment:
We're also still using this for CPU, correct? I assume that's to prevent changes to that file in this PR, and we'll eventually move it over to the new interface as well.

Reply:
Yes that's correct. I think I'll just remove this method and force the use of the new one. We've been making breaking changes to the interface lately anyway, so keeping this for BC was probably overkill.